core-api/src/main/java/com/optimizely/ab/odp/parser/impl/GsonParser.java

```diff
@@ -34,15 +34,16 @@ public List<String> parseQualifiedSegments(String responseJson) {
             JsonObject root = JsonParser.parseString(responseJson).getAsJsonObject();
 
             if (root.has("errors")) {
-                JsonArray errors =  root.getAsJsonArray("errors");
-                StringBuilder logMessage = new StringBuilder();
-                for (int i = 0; i < errors.size(); i++) {
-                    if (i > 0) {
-                        logMessage.append(", ");
+                JsonArray errors = root.getAsJsonArray("errors");
+                JsonObject extensions = errors.get(0).getAsJsonObject().get("extensions").getAsJsonObject();
+                if (extensions != null && extensions.has("code")) {
+                    if (extensions.get("code").getAsString().equals("INVALID_IDENTIFIER_EXCEPTION")) {
+                        logger.warn("Audience segments fetch failed (invalid identifier)");
+                    } else {
+                        String errorMessage = extensions.get("classification") == null ? "decode error" : extensions.get("classification").getAsString();
+                        logger.error("Audience segments fetch failed (" + errorMessage + ")");
```

Comment:
it's a common case that extensions has "classification" but no "code".
We should log error message for that too and re-use this code for that case.